File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    
    void solve(vector<int>& arr,int idx,int target,vector<int>& temp,vector<vector<int>>& ans){
        
        
        if(target==0){
            ans.push_back(temp);
            return;
        }
           
        
        for(int i=idx;i<arr.size();i++){
            
            if(arr[idx]>target) break; // if curr ele is bigger than target then all rest wil be bigger so no point in includingh
            
            if(i>idx and arr[i]==arr[i-1])
                continue;  // skip taking duplicate elements;
            
            // if not duplicate then include and try
            temp.push_back(arr[i]);
            solve(arr,i+1,target-arr[i],temp,ans);
            temp.pop_back();
        }
        
        
        
    }
    
    
    vector<vector<int>> combinationSum2(vector<int>& arr, int target) {
        
        sort(arr.begin(),arr.end());
        
        
        vector<vector<int>> ans;
        
        vector<int> temp;
        
        solve(arr,0,target,temp,ans);
                
        return ans;
        
    }
};

File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
    
    def solve(self,idx,target,nums,temp,ans):
        
        if(target==0):
            ans.append(list(temp))
        
        for i in range(idx,len(nums)):
            
            if(nums[i]>target):
                break
            
            if(idx!=i and nums[i]==nums[i-1]):
                continue
            
            temp.append(nums[i])
            self.solve(i+1,target-nums[i],nums,temp,ans)
            temp.pop()
            
        
        
        
    
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        idx = 0
        temp = []
        ans = []
        nums.sort()
        
        self.solve(idx,target,nums,temp,ans)
        return ans
        